cpp/daal/include/services/service_profiler.h

```diff
@@ -0,0 +1,468 @@
+{
+
+inline static void set_verbose_from_env()
+{
+    static std::mutex mutex;
+    std::lock_guard<std::mutex> lock(mutex);
+    const char * verbose_str = std::getenv("ONEDAL_VERBOSE");
+    int newval               = 0;
+    if (verbose_str)
+    {
+        char * endptr  = nullptr;
+        errno          = 0;
+        long val       = std::strtol(verbose_str, &endptr, 10);
+        bool parsed_ok = (errno == 0 && endptr != verbose_str && *endptr == '\0');
+        if (parsed_ok && val >= 0 && val <= 5) newval = static_cast<int>(val);
+    }
+    daal_verbose_val = newval;
+}
+
+inline static int daal_verbose_mode()
+{
+    if (daal_verbose_val == -1) set_verbose_from_env();
+    return daal_verbose_val;
+}
+
+inline static std::string format_time_for_output(std::uint64_t time_ns)
+{
+    std::ostringstream out;
+    double time = static_cast<double>(time_ns);
+    if (time <= 0)
+        out << "0.00s";
+    else if (time > 1e9)
+        out << std::fixed << std::setprecision(2) << time / 1e9 << "s";
+    else if (time > 1e6)
+        out << std::fixed << std::setprecision(2) << time / 1e6 << "ms";
+    else if (time > 1e3)
+        out << std::fixed << std::setprecision(2) << time / 1e3 << "us";
+    else
+        out << static_cast<std::uint64_t>(time) << "ns";
+    return out.str();
+}
+
+inline void print_header()
+{
+    daal::services::LibraryVersionInfo ver;
+    std::cerr << "Major version:          " << ver.majorVersion << '\n';
+    std::cerr << "Minor version:          " << ver.minorVersion << '\n';
+    std::cerr << "Update version:         " << ver.updateVersion << '\n';
+    std::cerr << "Product status:         " << ver.productStatus << '\n';
+    std::cerr << "Build:                  " << ver.build << '\n';
+    std::cerr << "Build revision:         " << ver.build_rev << '\n';
+    std::cerr << "Name:                   " << ver.name << '\n';
+    std::cerr << "Processor optimization: " << ver.processor << '\n';
+    std::cerr << '\n';
+}
+
+inline void profiler_log_named_args(const char * /*names*/) {}
+
+template <typename T, typename... Rest>
+inline void profiler_log_named_args(const char * names, const T & value, Rest &&... rest)
+{
+    const char * comma = strchr(names, ',');
+    std::string name   = comma ? std::string(names, comma) : std::string(names);
+    name.erase(name.begin(), std::find_if(name.begin(), name.end(), [](unsigned char ch) { return !std::isspace(ch); }));
+    std::cerr << name << ": " << value << "; ";
+    if (comma) profiler_log_named_args(comma + 1, std::forward<Rest>(rest)...);
+}
+
+inline static bool is_service_debug_enabled()
+{
+    static const bool service_debug_value = [] {
+        int value = daal_verbose_mode();
+        return value == 5;
+    }();
+    return service_debug_value;
+}
+
+inline static bool is_logger_enabled()
+{
+    static const bool logger_value = [] {
+        int value = daal_verbose_mode();
+        return value == 1 || value == 4 || value == 5;
```

Comment:
The magic constants `1`, `4`, `5`, etc. have to be named somehow.

Reply:
Fixed! Thanks!